src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java

```diff
@@ -167,8 +167,14 @@
         // 5. Fill in the reads. Each shard is an assembly region, with its overlapping reads.
         JavaRDD<Shard<GATKRead>> assemblyRegionShardedReads = SparkSharder.shard(ctx, reads, GATKRead.class, header.getSequenceDictionary(), assemblyRegionBoundaries, shardingArgs.readShardSize);
 
-        // 6. Convert shards to assembly regions.
-        JavaRDD<AssemblyRegion> assemblyRegions = assemblyRegionShardedReads.map((Function<Shard<GATKRead>, AssemblyRegion>) shard -> toAssemblyRegion(shard, header));
+        // 6. Convert shards to assembly regions. Reads downsampling is done again here, and is assumed to be consistent
+        // with the downsampling done in step 1, since it is deterministic by locus.
+        JavaRDD<AssemblyRegion> assemblyRegions = assemblyRegionShardedReads.mapPartitions((FlatMapFunction<Iterator<Shard<GATKRead>>, AssemblyRegion>) shardedReadIterator -> {
+            final ReadsDownsampler readsDownsampler = assemblyRegionArgs.maxReadsPerAlignmentStart > 0 ?
+                    new PositionalDownsampler(assemblyRegionArgs.maxReadsPerAlignmentStart, header) : null;
```

Comment:
When #5437 goes in, you want to build this positional downsampler with the deterministic arguments enabled.

Reply:
Right.